lib/puppet/defaults.rb

```diff
@@ -186,6 +186,11 @@ def self.valid_file_checksum_types
           Facter.trace(value) if Facter.respond_to? :trace
         end
     },
+    :puppet_trace => {
+        :default  => false,
```

Comment:
Does having this default to false mimic previous behavior, or maintain the performance improvement from the previous patch?

Reply:
It doesn't have an effect on either, actually. Right now the PR will interleave the ruby and puppet stack traces if the `--trace` flag is passed (rationale being the expensive part is the getting of the Ruby trace, adding the puppet trace is relatively inexpensive and doesn't dilute the signal to noise ratio much if at all), if the `--trace` flag isn't passed but the `--puppet_trace` flag _is_, then we will just output the puppet stack trace, which everyone thought was the most value for the majority of users (high signal to noise ratio for actual puppet authors). ie, using the reproducer in the ticket,
with `--trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions/fail.rb:3:in block in <top (required)>'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:174:in block (2 levels) in newfunction'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler/around_profiler.rb:58:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/util/profiler.rb:51:in profile'
/Users/justin/Backup/code/work/puppet/lib/puppet/parser/functions.rb:167:in block in newfunction'
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/lib/puppet/pops/puppet_stack.rb:41:in stack'
...

```





with just `--puppet_trace`:

```
Error: Evaluation Error: Error while evaluating a Function Call, somethinng (file: /Users/justin/Backup/code/work/puppet/broken.pp, line: 2, column: 3) on node localhost
/Users/justin/Backup/code/work/puppet/broken.pp:2
/Users/justin/Backup/code/work/puppet/broken.pp:6
/Users/justin/Backup/code/work/puppet/broken.pp:9
/Users/justin/Backup/code/work/puppet/broken.pp:0
```

Reply:
Ahh I get it, thanks for clearing that up.

Reply:
Think you could add something to the doc string for `trace` noting that it will contain both the Ruby and the Puppet frames? I think that would help disambiguate these two.

Reply:
Updated, let me know if it's still lacking.

Reply:
What happens if both `--trace` and `--puppet_trace` are passed?

Reply:
`--puppet_trace` is a noop then, I can add a comment about that as well!

Reply:
updated